interfaces/symlinks/backend.go

```diff
@@ -0,0 +1,175 @@
+		return fmt.Errorf("cannot obtain symlinks specification for snap %q: %s",
+			snapName, err)
+	}
+
+	return b.ensureSymlinks(spec.(*Specification), symlinkDirs)
+}
+
+// Remove removes modules symlinks files specific to a given snap.
+// This method should be called after removing a snap.
+func (b *Backend) Remove(snapName string) error {
+	// If called for the system (snapd) snap, that is possible only in a
+	// classic scenario when all other snaps in the system must have been
+	// removed already to allow the removal of the snapd snap. In that
+	// case, the config files will have already been removed by a Setup
+	// call, so we do not need to do anything here.
+
+	// TODO but this needs to be revisited for when we start supporting
+	// symlinks plugs in snaps.
+	return nil
+}
+
+// NewSpecification returns a new specification associated with this backend.
+func (b *Backend) NewSpecification(*interfaces.SnapAppSet,
+	interfaces.ConfinementOptions) interfaces.Specification {
+	return &Specification{}
+}
+
+// SandboxFeatures returns the list of features supported by snapd for symlinks.
+func (b *Backend) SandboxFeatures() []string {
+	return []string{"mediated-symlinks"}
+}
+
+func (b *Backend) ensureSymlinks(spec *Specification, symlinkDirs map[string]bool) error {
+	// Setup symlinks only if the snap has plugs that require it. For the
+	// moment this is only the system snap.
+	if len(spec.plugs) == 0 {
+		return nil
+	}
+
+	for lnDir := range spec.dirsToLinkToTarget {
+		if _, ok := symlinkDirs[lnDir]; !ok {
+			return fmt.Errorf("internal error: %s not in any registered symlinks directory", lnDir)
+		}
+	}
+
+	// TODO supported directories apply currently only to a classic rootfs,
+	// not to the rootfs of a snap.
+	for dir := range symlinkDirs {
+		entries, err := os.ReadDir(dir)
+		dirExists := true
+		if err != nil {
+			if _, ok := err.(*os.PathError); ok {
+				dirExists = false
+			} else {
+				return err
+			}
+		}
+		lnsToTarget := spec.dirsToLinkToTarget[dir]
+		// Loop to remove now unwanted symlinks
+		for _, node := range entries {
+			if node.Type() != fs.ModeSymlink {
+				continue
+			}
+			// these symlinks are absolute
+			path := filepath.Join(dir, node.Name())
+			dest, err := os.Readlink(path)
+			if err != nil {
+				return err
+			}
+			// path is under $SNAP or part of snaps data, otherwise
+			// it is not under snapd control.
+			//
+			// TODO an alternative would be to add trusted extended
+			// attributes to these symlinks (xattr(7)) - adding
+			// with the equivalent of
+			// "sudo setfattr -h -n trusted.<snap> -v <val> <symlink>".
+			if !strings.HasPrefix(dest, dirs.SnapMountDir+"/") &&
+				!strings.HasPrefix(dest, dirs.SnapDataDir+"/") {
+				continue
+			}
+			// link is active
+			if _, ok := lnsToTarget[node.Name()]; ok {
+				continue
+			}
+
+			// link not active, remove
+			if err := os.Remove(path); err != nil {
+				logger.Noticef("symlinks backend cannot remove %q", path)
+			}
```

Comment:
this logic about preexisting symlinks either active or to remove does not seem tested atm

Reply:
Good catch, there was an issue with the tests. Interestingly this has lead to some changes in testutil, as the checker sort of unexpectedly follows symlinks. I've defined new checkers that do not do this.